src/connectiondataholder.cpp

```diff
@@ -216,50 +217,44 @@ void ConnectionDataHolder::updateIpAddress() {
   }
   m_updatingIpAddress = true;
 
-  NetworkRequest* request = NetworkRequest::createForIpInfo(this);
-  connect(request, &NetworkRequest::requestFailed,
-          [this](QNetworkReply::NetworkError error, const QByteArray&) {
-            logger.log() << "IP address request failed" << error;
-
-            ErrorHandler::ErrorType errorType =
-                ErrorHandler::toErrorType(error);
-            if (errorType == ErrorHandler::AuthenticationError) {
-              MozillaVPN::instance()->errorHandle(errorType);
-            }
-
-            m_updatingIpAddress = false;
-            emit ipAddressChecked();
-          });
-
-  connect(request, &NetworkRequest::requestCompleted,
-          [this](const QByteArray& data) {
-            logger.log() << "IP address request completed";
-            QJsonDocument json = QJsonDocument::fromJson(data);
-            if (json.isObject()) {
-              QJsonObject obj = json.object();
-
-              QJsonValue country = obj.take("country");
-              if (m_checkStatusTimer.isActive() &&
-                  country.toString().toLower() !=
-                      MozillaVPN::instance()->currentServer()->countryCode()) {
-                // In case the country-we're reported in does not match the
-                // connected server we may retry only once.
-                logger.log() << "Reported ip not in the right country, retry!";
-                TimerSingleShot::create(this, 3000,
-                                        [this]() { updateIpAddress(); });
-              }
-              QJsonValue value = obj.value("ip");
-              if (value.isString()) {
-                m_ipAddress = value.toString();
-                logger.log() << "Set own Address: " << m_ipAddress << " in "
-                             << country.toString();
-                emit ipAddressChanged();
-              }
-            }
-
-            m_updatingIpAddress = false;
-            emit ipAddressChecked();
-          });
+  IPFinder* ipfinder = new IPFinder(this);
+  connect(
+      ipfinder, &IPFinder::completed,
+      [this](const QString& ipv4, const QString& ipv6, const QString& country) {
+        if (ipv4.isEmpty() && ipv6.isEmpty()) {
+          logger.log() << "IP address request failed";
+          m_updatingIpAddress = false;
+          emit ipAddressChecked();
+          return;
+        }
+
+        logger.log() << "IP address request completed";
+        if (m_checkStatusTimer.isActive() &&
+            country != MozillaVPN::instance()->currentServer()->countryCode()) {
+          // In case the country-we're reported in does not match the
+          // connected server we may retry only once.
+          logger.log() << "Reported ip not in the right country, retry!";
+          TimerSingleShot::create(this, 3000, [this]() { updateIpAddress(); });
```

Comment:
I don't see the mechanism here that would prevent this from looping forever every 3000ms in the event that the country code never matches.

Reply:
This is a good point. But I haven't touched this part. Can we work on it as a follow up?